### ISSUE TITLE
httpcli should support HTTP 1.1

### ISSUE THREAD

**Issue by kalyanac** [grpc/grpc#19173]
httpcli [uses](https://github.com/grpc/grpc/blob/master/src/core/lib/http/format_request.cc#L36) HTTP/1.0. 
Istio Envoy requires HTTP/1.1 or above. HTTP 1.0 support is available in Istio 1.1 ([PR](https://github.com/istio/istio/issues/11235)), which is still in [alpha](https://cloud.google.com/istio/docs/istio-on-gke/installing) on GKE. On Istio 1.1, users still need to set an environment HTTP10 to enable HTTP/1.0 support.

When using grpc httpcli to contact the metadata server, requests are intercepted by Envoy proxy which returns HTTP 426 Upgrade required for any HTTP/1.0 requests.
 
### What version of gRPC and what language are you using?
 1.21
 
### What operating system (Linux, Windows,...) and version?
 Linux, Ubuntu 14.04
 
### What did you do?
httpcli request to "http://metadata.google.internal.//computeMetadata/v1/instance/service-accounts/default/token" with required "Metadata-Flavor: Google" header
 
### What did you expect to see?
 HTTP status 200 and a token
 
### What did you see instead?
HTTP Status 426 upgrade required
 
Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).
Patched [format_request.cc](https://github.com/grpc/grpc/blob/master/src/core/lib/http/format_request.cc#L36) to 1.1 and the request succeeds.

See [TROUBLESHOOTING.md](https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md) for how to diagnose problems better.
 
### Anything else we should know about your project / environment?
https://github.com/GoogleCloudPlatform/cloud-profiler-java
and other projects where grpc httpcli is used to contact the GCP metadata server.

**Comment by markdroth** [comment]
@kalyanac, can you provide more context about this?  The httpcli code is not part of grpc's public API; it's there purely for our own internal use, and we don't support it for use in other applications.  Given that, I'm curious how you ran into this problem.

**Comment by kalyanac** [comment]
We do not use http_cli directly. We use gRPC and that seems to internally use http_cli to get a token from
"http://metadata.google.internal.//computeMetadata/v1/instance/service-accounts/default/token"

We noticed this from the Istio proxy logs. 
Please see [b/126601115#comment10](b/126601115#comment10) for screenshots. This can be easily reproduced.

**Comment by markdroth** [comment]
Ah, okay -- yes, that's done by the gRPC oauth credentials.

I agree that in principle, we should change our HTTP client to support HTTP/1.1 (and maybe even HTTP/2), and I'll keep this issue open to track that work.  However, our team is currently quite busy with other work, and since this has a work-around (you mentioned the HTTP/1.0 support in Istio 1.1), we are unlikely to prioritize this in the near future.

**Comment by kalyanac** [comment]
Thank you for triaging this. 
Given the importance of Istio and the popularity of gRPC, I suspect this may be affecting others as well.

**Comment by kalyanac** [comment]
Any update on this? I was able to workaround the problem by enabling HTTP 1.0 support in Pilot.

**Comment by markdroth** [comment]
No update.  We're still not prioritizing this, since there's a usable work-around, and we have a lot of other higher-priority work.

**Comment by kalyanac** [comment]
Any update on this?

**Comment by markdroth** [comment]
No, no change.  We've got too many other higher priorities, so this probably isn't going to get any attention unless someone wants to contribute some code

**Comment by gusior** [comment]
Hey,
FYI, I just encountered the very same issue when testing out the Google Cloud Platform Anthos Service Mesh setup + GKE + GKE Workload Identity setup at my company. grpc httpcli is being used to obtain service account token, but it fails due to HTTP/1.0 being used which is not supported by Istio.

**Comment by SHiftLin** [comment]
I also encountered the same problem when I used envoy as a proxy. I hope this could be fixed soon

**Comment by nicolasnoble** [comment]
Ah, it's likely we'll also need a chunked encoding parser.

**Comment by ctiller** [comment]
We *may* want to consider keeping the connection open for multiple requests once we do this.
